### PR TITLE
i2c: Fujitsu FRAM based I2C Bus driver testing

### PR DESCRIPTION
**Pull request by teburd**
~~Ignore first commit, twister fails to run without it for me atm.~~ See #65995

The current test for I2C driver testing is based on a "gy271" which I don't have on my desk for one thing.

There's a variety of RAM like i2c parts (FRAM, SRAM with flash write on power out, etc).

Adds a second test, and includes the callback and RTIO APIs being covered using RAM like reads and writes to an external memory device over i2c. Tested with fujistu mb85 fram module.

In draft until I work out some things, e.g. how to choose which addressing scheme (1 or 2 byte memory addressing) and such.

Sadly my initial plan to use the fram click 3 and arduino fram breakouts didn't quite pan out as the fram click 3 uses a mb94 which is a bit more than a simple fram part and requires sending formatted commands rather than simple read/write memory addresses.

My new plan is a general RAM test where the external device needs to respond to 2 byte memory address read and write requests just like the mb85, there just happens to be a few parts that work like that besides the mb85 that also have mikrobus breakouts.

### CODE REVIEW

**Comment by hakehuang** [comment]
@teburd, as we discuss in test working groupd, we can use https://www.mikroe.com/fram-3-click as candidate. and  as not all  evk boards has such FRAM on board, it is also recommand that add this bridge card for https://www.mikroe.com/arduino-uno-click-shield, so that click interface can goes with arduino interfaces. Do you agree?

File: tests/drivers/i2c/i2c_fram/testcase.yaml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+tests:
+  drivers.i2c.api_fram:
+    depends_on: i2c
+    tags:
+      - drivers
+      - i2c
+    filter: dt_alias_exists("mb85")
```
Comment (hakehuang) [inline tests/drivers/i2c/i2c_fram/testcase.yaml:7]: we need add harness requirement for click_mb85

Comment (teburd) [reply tests/drivers/i2c/i2c_fram/testcase.yaml:7]: I think leaving the fixture as mb85 is ok, there *are* other breakouts, the key here is that the board has a i2c bus alias labeled mb85, that could be the i2c bus connected to a particular click board socket.

Speaking of click boards, it seems like we might want an added alias for click_i2cX as there's the strong potential, e.g. the nucleo 144 shield, where more than one i2c is possible.

I don't know the best route to deal with multiple shield boards though and multiple shield board sockets, maybe @jfischer-no would have some helpful hints on that front.

**Comment by teburd** [comment]
> @teburd, as we discuss in test working groupd, we can use https://www.mikroe.com/fram-3-click as candidate. and as not all evk boards has such FRAM on board, it is also recommand that add this bridge card for https://www.mikroe.com/arduino-uno-click-shield, so that click interface can goes with arduino interfaces. what is your testing setting?

I have those on order, I had the adafruit breakout (i2c fram https://www.adafruit.com/product/1895) which I think I might've dorked up as I kept moving the wires around. Exactly why I like the sockets and board to board interfaces better...

File: tests/drivers/i2c/i2c_fram/testcase.yaml (hunk 1 above)
Comment (hakehuang) [reply tests/drivers/i2c/i2c_fram/testcase.yaml:7]: may just put `arduino`  will be ok

**Comment by teburd** [comment]
I did some adhoc runs with this test and... already finding bugs it seems

| Board                | I2C Transfer | I2C Write Read | I2C Transfer Callback | RTIO        |
| -------------------- | ------------ | -------------- | --------------------- | ----------- |
| frdm_k64f            | pass         | pass           | fail                  | unsupported |
| mimxrt1010_evk       | fail         | fail           | fail                  | unsupported |
| nrf52840dk_nrf52840  | pass         | pass           | fail                  | unsupported |

**Comment by hakehuang** [comment]
#68400 , I propse to switch such apparch for all peripheral cases

**Comment by teburd** [comment]
I'd note the uno click shield uses the wrong pins for our arduino shield i2c and instead one of the frdm click shields should be used. https://www.mikroe.com/frdm-kl46-click-shield or https://www.mikroe.com/frdm-k64-click-shield

The i2c pins for the kl46 shield connect to the arduino_i2c aliased i2c pins as would be expected for both clicks, the k64 click shield would work only with one slot for i2c with our arduino_i2c setup today.

File: tests/drivers/i2c/i2c_fram/testcase.yaml (hunk 1 above)
Comment (jfischer-no) [reply tests/drivers/i2c/i2c_fram/testcase.yaml:7]: @teburd sorry, not sure I understand, maybe `boards/shields/arduino_uno_click` is an example of what you are looking for.

Comment (teburd) [reply tests/drivers/i2c/i2c_fram/testcase.yaml:7]: I'll take a closer look at that, thank you

**Comment by teburd** [comment]
Was merged in as part of #68776

**Pull request closed without merging**